Add stats api.

adnapibar opened the pull request:
Issue #302 

The `/_ltr/stats` api provides basic information on the plugin usage and performance, including,
- overall status of the plugin health.
- statistics on the feature stores.
- statistics on the internal caches used by the plugin.

nathancday commented:
Thanks for building this out and contributing it! I'm checking it out locally now and I'm sure this will be useful to a lot of plugin users.

@nomoa would you review this PR when you have a chance?

nathancday commented:
Also @ebernhardson if you want to take a look, I'm interested if this would work to resolve the issue you raised about cache churn in #140

nathancday commented:
Hey @adnapibar could you do a quick brain dump to about how you designed the stats api to work and add a new section for its usage to `docs/advanced-functionality.rst`?

adnapibar commented:
> 
> 
> Looks great, thanks!
> 
> I haven't fully reviewed the patch yet but I wanted to send a first round of comments.
> 
> The main concern I have is the blocking call in StoreUtils on the Client to get the counts of the various elements stored. I think this is problematic to have a blocking call there and we should flip the logic to use the async APIs like everything else. But overall what StoreUtils is extracting should probably be clarified a bit more.
> 
> One thing that would be very nice is implementing some tests in `src/test/resources/rest-api-spec/`.
> 
> Thanks!

Thanks @nomoa for your review. I couldn't get to resolve the comments earlier, sorry about that. I have resolved most of your earlier comments. I agree the StoreUtils was not abstracting much, so I have removed that and moved the code to the suppliers.  The code to get the counts is now replaced with an aggregate query (not sure if that resolves the blocking concern that you have, any pointers on that would be great!).  Also, added some tests in the rest-api-specs. 

Thanks!

nomoa commented on src/main/java/com/o19s/es/ltr/stats/suppliers/StoreStatsSupplier.java:
```diff
@@ -0,0 +1,138 @@
+    private void updateCount(Terms.Bucket bucket, Map<String, Object> storeStat) {
+        storeStat.computeIfPresent(
+                typeToStatName(bucket.getKeyAsString()),
+                (k, v) -> bucket.getDocCount());
```
should you add v? `bucket.getDocCount() + v`
otherwise you might only get the numbers for the last response of you msearch request.

adnapibar replied:
Hi @nomoa thanks for the review. I have update this in the latest commit. 

However this should not be a bug IMO. The aggregations will return one term bucket for each type of element, Isn't that correct? In that case, it won't be called more than once per element per store (each store has separate counts map). The initial count for each of these element is 0 and updateCount method is updating this to the actual number of elements in the store. 

Please let me know if I have misunderstood this. I have however updated the code to include

nomoa commented on src/main/java/com/o19s/es/ltr/stats/suppliers/StoreStatsSupplier.java:
```diff
@@ -0,0 +1,138 @@
+                        clusterService.state().metadata().index(storeName),
+                        clusterService.state().getRoutingTable().index(storeName));
+
+        return indexHealth.getStatus().name().toLowerCase(Locale.getDefault());
```
nit: I think it's preferable to use Locale.ROOT to not rely on system defaults

nomoa commented on src/main/java/com/o19s/es/ltr/stats/suppliers/StoreStatsSupplier.java:
```diff
@@ -0,0 +1,138 @@
+        try {
+            MultiSearchResponse msr = requestBuilder.execute().get();
+            assert indices.size() == msr.getResponses().length;
+            Map<String, Map<String, Object>> stats = new ConcurrentHashMap<>(indices.size());
```
I don't think you need a concurrent hashmap here, if you need one I probably missed something in your PR.

nomoa commented:
Forgot to mention the blocking calls:
I think it's OK to leave them, it is generally not great to call `get()` on client requests, reason is that you hold 2 threads from the threadpool for the operation: one waiting and another one doing the job.

Just for reference the way to fix this is to switch to async calls using the `execute` method with an `ActionListener` but this seems hard in the current state of the PR since it relies on Supplier::get() to gather its metrics.

nathancday commented:
Thanks again @adnapibar for all of the new functionality and @nomoa for all of the review. I just requested a second set of @worleydl

softwaredoug commented on src/test/resources/rest-api-spec/test/fstore/90_get_stats.yml:
```diff
@@ -0,0 +1,174 @@
+  - do:
+      ltr.get_stats:
+        stat: "status"
+  - match: { status: "green" }
```
Impressive YAML programming ;)

adnapibar replied:
Thanks

adnapibar commented on src/main/java/com/o19s/es/ltr/stats/LTRStat.java:
```diff
@@ -0,0 +1,39 @@
+        return clusterLevel;
+    }
+
+    public T getValue() {
```
Updated this to `getStatValue`.

adnapibar commented on src/main/java/com/o19s/es/ltr/stats/LTRStat.java:
```diff
@@ -0,0 +1,39 @@
+        return supplier.get();
+    }
+
+    public void increment() {
```
I have removed `CounterSupplier`. I added this initially as I was thinking to later add stats providing information on number of success and failed requests to the plugin, and this counter will be helpful.  I'll add it in that PR later.

adnapibar commented on src/main/java/com/o19s/es/ltr/stats/StatName.java:
```diff
@@ -0,0 +1,44 @@
+    STORES("stores"),
+    CACHE("cache"),
+
+    STORE_STATUS("status"),
```
Thanks, I have moved those to the individual stat suppliers.

adnapibar commented on src/main/java/com/o19s/es/ltr/stats/suppliers/PluginHealthStatusSupplier.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    private String combineStatuses(String s1, String s2) {
+        if (s2 == null || STATUS_RED.equals(s1) || STATUS_RED.equals(s2)) {
```
Agree, the check seems unnecessary. I have removed it.

adnapibar commented on src/main/java/com/o19s/es/ltr/stats/suppliers/StoreStatsSupplier.java:
```diff
@@ -0,0 +1,138 @@
+                .forEach(idx -> {
+                    indices.add(idx);
+                    requestBuilder.add(countSearchRequest(idx));
+                });
```
Thanks!

adnapibar commented on docs/advanced-functionality.rst:
```diff
@@ -211,4 +211,66 @@ If (and only if) the extra logging Map is accessed, it will be returned as an ad
+=============================
+The stats API gives the overall plugin status and statistics::
+
+    GET /_ltr/stats
```
Thanks!, updated.

adnapibar commented on src/main/java/com/o19s/es/ltr/action/TransportLTRStatsAction.java:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    @Override
+    protected void doExecute(Task task, LTRStatsNodesRequest request, ActionListener<LTRStatsNodesResponse> listener) {
```
Removed.

adnapibar commented on src/main/java/com/o19s/es/ltr/stats/LTRStat.java:
```diff
@@ -0,0 +1,39 @@
+ *
+ * @param <T> the type of the value returned by the stat.
+ */
+public class LTRStat<T> {
```
Updated.

adnapibar commented:
Thanks @nomoa and @softwaredoug  for taking time to review my PR. I have addressed your comments and made the requested changes. Please let me know if there is anything else that I have missed or should add.

nomoa commented on src/main/java/com/o19s/es/ltr/stats/suppliers/StoreStatsSupplier.java:
```diff
@@ -0,0 +1,153 @@
+                    aggs.getBuckets()
+                            .stream()
+                            .filter(Objects::nonNull)
+                            .forEach(bucket -> updateCount(bucket, storeStat));
```
sorry my bad, I misread the scope of storeStat in that case I think it'd make more sense to collect and then merge, e.g.:
```
                    aggs.getBuckets()
                            .stream()
                            .filter(Objects::nonNull)
                            .collect(Collectors.toMap(Terms.Bucket::getKeyAsString, Terms.Bucket::getDocCount))
                            .forEach(storeStat::put);
```

adnapibar commented on src/main/java/com/o19s/es/ltr/action/LTRStatsAction.java:
```diff
@@ -0,0 +1,180 @@
+        }
+    }
+
+    public static class LTRStatsNodesRequest extends BaseNodesRequest<LTRStatsNodesRequest> {
```
I kept it as per existing conventions in the repo. One alternative convention might be to use `LTRStatsNodeRequest` for single node and `LTRStatsRequest` for nodes. ([example](https://github.com/opendistro-for-elasticsearch/anomaly-detection/tree/master/src/main/java/com/amazon/opendistroforelasticsearch/ad/transport))

nathancday commented:
@softwaredoug and @nomoa green-light for merge?

nomoa commented:
@nathancday LGTM

adnapibar commented:
@nathancday Can we merge it now?

worleydl commented:
Merging now.  Should see a release next week.
